[quant][graph] Add pass to insert quant dequant for dynamic quantization

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35586 [quant][graph] Add quant fusion for dynamic quantization
* **#35448 [quant][graph] Add pass to insert quant dequant for dynamic quantization**

Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755679](https://our.internmc.facebook.com/intern/diff/D20755679)

supriyar committed:
[quant][graph] Add pass to insert quant dequant for dynamic quantization
Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -630,6 +630,29 @@ graph(%self, %a, %b, %inplace):
   };
 };
 
+std::vector<Node*> getCallMethods(Value* v) {
```
this is not getting call method for value `v`, but getting all call method in the graph.

to get call method for v, you just need to look at uses of v and check if the node is a `CallMethod` or not

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1689,6 +1770,7 @@ void InsertQuantDeQuantHelper::run(
   GRAPH_DUMP("Before Quantize Tensors:", graph);
   Value* self = graph->inputs()[0];
   quantizeTensors(module, graph.get(), self);
+
```
remove?

jerryzh168 reviewed (changes requested):
the logic checking if a value is weight still needs some cleanup, other code looks fine

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1277,26 +1328,49 @@ void insertQuantDeQuantCall(
+  if (is_dynamic && !isWeightOfConvOrLinear(v)) {
+    std::string choose_qparams_func = "_choose_qparams_per_tensor";
+    auto reduce_range = g->insertConstant(false);
+    choose_qparams = g->create(at::Symbol::aten(choose_qparams_func), v, 2);
```
what is `2`?

jerryzh168 replied:
looks like this is number of outputs. can you add a comment?
also can you put all inputs together?
like 
`g->create(at::Symbol::aten(choose_qparams_func), {v, reduce_range}, 2);`

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1277,26 +1328,49 @@ void insertQuantDeQuantCall(
+    choose_qparams = g->create(at::Symbol::aten(choose_qparams_func), v, 2);
+    choose_qparams->addInput(reduce_range);
+
+    choose_qparams->output(0)->setDebugName(v->debugName() + ".scale_");
```
nit: `.scale_` -> `.scale`?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1277,26 +1328,49 @@ void insertQuantDeQuantCall(
+    quant = g->create(at::Symbol::aten(quantize_func), quant_inputs);
+    g->insertNode(quant);
+  } else {
+    quant = g->create(at::Symbol::aten(quantize_func), inputs);
```
could you move the construction of `inputs` to this branch?

supriyar replied:
I'm using the `dtype` information from the inputs in the `quant_inputs`

jerryzh168 replied:
you can still move that inside this branch and create a getattr node for dtype for `quant_inputs`, this makes me think that we probably want to change `qparam_names` to be a map as well, but that can be a different PR

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1277,26 +1301,50 @@ void insertQuantDeQuantCall(
+    std::string choose_qparams_func = "_choose_qparams_per_tensor";
+    auto reduce_range = g->insertConstant(false);
+    // choose_qparams_per_tensor has 2 outputs, (scale, zero_point).
+    choose_qparams = g->create(at::Symbol::aten(choose_qparams_func), {v, reduce_range}, 2);
```
nit: please add a comment for 2 as well:
g->create(at::Symbol::aten(choose_qparams_func), {v, reduce_range}, /* num_outputs = */ 2);

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -694,6 +694,29 @@ bool isWeightOfConvOrLinear(Value* v) {
+    Node* n = u.user;
+    if (n->kind() == prim::CallMethod) {
+      auto m = getInvokedModule(module, n, self);
+      std::shared_ptr<Graph> g = m.get_method(n->s(attr::name)).graph();
```
nit: you can do `auto` here

jerryzh168 replied:
can you change the type to `auto` here?

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -694,6 +694,29 @@ bool isWeightOfConvOrLinear(Value* v) {
+      std::shared_ptr<Graph> g = m.get_method(n->s(attr::name)).graph();
+
+      for (int i = 0; i < n->inputs().size(); ++i) {
+        if (n->inputs()[i] == v) {
```
you can use `v->offset()` to get the offset of `v`

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -694,6 +694,29 @@ bool isWeightOfConvOrLinear(Value* v) {
+  if (isWeightOfConvOrLinear(v)) {
+    return true;
+  }
+  bool result = false;
```
I think we want also make sure that all uses give the same result for this check, you can use a similar pattern as: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization.cpp#L983 to do that

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -694,6 +694,29 @@ bool isWeightOfConvOrLinear(Value* v) {
+    return true;
+  }
+  bool result = false;
+  auto* self = graph->inputs()[0]; 
```
this can be `v->owningGraph()->inputs()[0]` then you don't need to pass around `graph`

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -694,6 +694,29 @@ bool isWeightOfConvOrLinear(Value* v) {
+  return result;
+}
+
+bool checkValueIsWeight(Module& module, Graph* graph, Value* v) {
```
can we have a better name? maybe just `isWeightOfConvOrLinear` and a comment that this is going through CallMethod

jerryzh168 replied:
or just `isWeight`

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1270,42 +1295,70 @@ void insertQuantDeQuantCall(
+    // Last argument is dtype.
+    auto dtype = g->insertGetAttr(self, qparam_names.back());
+    quant_inputs.push_back(dtype);
+    quant = g->create(at::Symbol::aten(quantize_func), quant_inputs);
```
nit: please set debug name for the output out this quant node as well

supriyar replied:
I've set the debug name in the line below this. Is there something else I need to set?

jerryzh168 replied:
no this is resolved

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -71,5 +72,101 @@ def forward(self, x):
+        qconfig_dict = {
+            '': script_qconfig(qconfig)
+        }
+        m = wrap_cpp_module(torch._C._jit_pass_insert_observers(m._c, "forward", qconfig_dict, False, True))
```
could you use prepare_dynamic_script here? also I added support for None in qconfig_dict, could you make this a function: https://github.com/pytorch/pytorch/blob/master/torch/quantization/_quantize_script.py#L76 and call it in prepare_dynamic_script as well?

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -71,5 +72,101 @@ def forward(self, x):
+
+        get_forward(m._c)(data)
+
+        m = wrap_cpp_module(torch._C._jit_pass_insert_quant_dequant(m._c, "forward", False, True))
```
same for this one, `convert_script`

supriyar replied:
I've added `convert_dynamic_script` in the next PR so I will update it there.

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -71,5 +72,101 @@ def forward(self, x):
+        qconfig_dict = {
+            '': script_qconfig(qconfig)
+        }
+        m = wrap_cpp_module(torch._C._jit_pass_insert_observers(m._c, "forward", qconfig_dict, False, True))
```
same here

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -71,5 +72,101 @@ def forward(self, x):
+
+        get_forward(m._c)(data)
+
+        m = wrap_cpp_module(torch._C._jit_pass_insert_quant_dequant(m._c, "forward", False, True))
```
same here

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -71,5 +72,101 @@ def forward(self, x):
                    .run(m.graph)
 
+
+    def test_insert_quant_dequant_dynamic(self):
```
rename this to `test_insert_quant_dequant_dynamic_conv`? since you have dynamic linear as well

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -71,5 +72,101 @@ def forward(self, x):
+                   .check("return") \
+                   .run(str(get_module_method(m, 'conv', '_conv_forward').graph))
+
+    def test_insert_quant_dequant_linear_dynamic(self):
```
dynamic_linear?

supriyar replied:
The quantized op is called `qlinear_dynamic` so I went with linear_dynamic instead of dynamic_linear.

jerryzh168 replied:
I see, OK

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -695,6 +695,32 @@ bool isWeightOfConvOrLinear(Value* v) {
+        }
+      } else {
+        result = call_method_result;
+      }
```
I don't think this works, what if the first CallMethod returns false, and second CallMethod returns true? you'll need to have an `optional<bool>` here I think

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1270,42 +1295,70 @@ void insertQuantDeQuantCall(
     bool is_per_channel,
-    const std::vector<std::string>& qparam_names) {
+    const std::vector<std::string>& qparam_names,
+    Module& module,
```
could you make `module` the first argument? it feels weird to put it here

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1399,8 +1453,14 @@ class InsertQuantDeQuantHelper {
+  bool is_dynamic = false;
 };
 
+void InsertQuantDeQuantHelper::setDynamicFlag(bool is_dynamic_) {
```
nit: you can move this to where it is declared since this is simple

jerryzh168 replied:
I think you want to rename the member `is_dynamic` to `is_dynamic_` to indicate that it is a member variable

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -752,6 +752,31 @@ bool isWeightOfConvOrLinear(Value* v) {
+      auto m = getInvokedModule(module, n, self);
+      auto g = m.get_method(n->s(attr::name)).graph();
+      auto call_method_result = isWeight(m, g->inputs()[u.offset]);
+      if (result) {
```
I think this has to be `result.has_value()`, since if result is false this might not hold either

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -752,6 +752,31 @@ bool isWeightOfConvOrLinear(Value* v) {
 }
 
+// Go through the CallMethod graph to check if the value is Weight.
+c10::optional<bool> isWeight(Module& module, Value* v) {
```
isWeight doesn't have to return optional

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -752,6 +752,31 @@ bool isWeightOfConvOrLinear(Value* v) {
+  if (isWeightOfConvOrLinear(v)) {
+    return true;
+  }
+  c10::optional<bool> result;
```
so, the optional just means we haven't seen any value yet

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -752,6 +752,31 @@ bool isWeightOfConvOrLinear(Value* v) {
+      if (result) {
+        // Check to make sure all the CallMethods in the graph produce the same output.
+        TORCH_CHECK(*call_method_result == *result,
+                      "Expected all CallMethods in the graph to produce same result.")
```
can you have a better error message e.g. `"Expect the value either to be non-weight or weight:", v->debugName()`

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -752,6 +752,31 @@ bool isWeightOfConvOrLinear(Value* v) {
+      } else {
+        result = call_method_result;
+      }
+    }
```
you need to check for the case when the node is not CallMethod as well I think

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -752,6 +752,32 @@ bool isWeightOfConvOrLinear(Value* v) {
+      if (result.has_value()) {
+        // Check to make sure all the CallMethods in the graph produce the same output.
+        TORCH_CHECK(call_method_result == result.value(),
+                    "Expected all CallMethods to produce either weight "
```
this is used by CallMethod, not produced by CallMethod..

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -70,5 +71,95 @@ def forward(self, x):
+        m = prepare_dynamic_script(m, {'': default_dynamic_qconfig})
+        data = torch.randn(1, 3, 10, 10, dtype=torch.float)
+
+        get_forward(m._c)(data)
```
I think we can do `m(data)` directly here

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -70,5 +71,95 @@ def forward(self, x):
+        assert len(m._modules._c.items()) == 1, \
+            'Expected to have single submodule of conv'
+
+        get_forward(m._c)(data)
```
same here

jerryzh168 reviewed (approved):
Please address some final comments

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add pass to insert quant dequant for dynamic quantization"


Summary:
Add _choose_qparams_per_tensor which returns scale and zero_point similar to the dynamic quantization in the operator

Test Plan:
python test/test_quantize_script.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755679](https://our.internmc.facebook.com/intern/diff/D20755679)

[ghstack-poisoned]
